File: curvyblocks-solution.py
import math
from decimal import *
while(True):
  try:
    curve1 = [Decimal(i) for i in input().split()]
    if len(curve1) == 0:
      break
    curve2 = [Decimal(i) for i in input().split()]
  except:
    break
  finalCurve = []
  for i in range(len(curve1)):
    finalCurve.append(curve2[i] - curve1[i])
  d = [finalCurve[1], finalCurve[2] * 2, finalCurve[3] * 3]

  discrim = (d[1] * d[1]) - (4 * d[2] * d[0])

  b = []
  solution = []
  try:
    if not d[2] == 0:
      if discrim >= 0:
        b.append((-d[1] + Decimal(math.sqrt(discrim)))/ (2 * d[2]))
        b.append((-d[1] - Decimal(math.sqrt(discrim)))/ (2 * d[2]))
      for s in b:
        if s < 1 and s > 0:
          solution.append(finalCurve[0] + (finalCurve[1] * s) + (finalCurve[2] * Decimal(math.pow(s, 2))) + finalCurve[3] * Decimal(math.pow(s,3)))
    else:
      if not d[1] == 0:
        s = (-d[0]/d[1])
        if s < 1 and s > 0:
            solution.append(finalCurve[0] + (finalCurve[1] * s) + (finalCurve[2] * Decimal(math.pow(s, 2))) + finalCurve[3] * Decimal(math.pow(s,3)))
  except:
    pass
  solution.append(finalCurve[0])
  solution.append(finalCurve[0] + finalCurve[1] + finalCurve[2] + finalCurve[3])
  small = min(solution)
  big = max(solution)
  print(big-small)

File: curvyblocks-solution.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Stacking_Curvy_Blocks
{
    class Program
    {
        static void Main(string[] args)
        {
            while (true)
            {
                double[] curve1;
                try
                {
                    curve1 = Console.ReadLine().Split().Select(x => double.Parse(x)).ToArray();
                }
                catch
                {
                    break;
                }
                if (curve1.Length == 0) break;

                double[] curve2 = Console.ReadLine().Split().Select(x => double.Parse(x)).ToArray();

                List<double> finalCurve = new List<double>();

                for(int i = 0; i < curve1.Length; i++)
                    finalCurve.Add(curve2[i] - curve1[i]);

                double[] derive = { finalCurve[1], finalCurve[2] * 2, finalCurve[3] * 3 };

                double discrim = Math.Pow(derive[1], 2) - (4 * derive[2] * derive[0]);

                List<double> b = new List<double>();
                List<double> sol = new List<double>();

                if(derive[2] != 0)
                {
                    if(discrim >= 0)
                    {
                        b.Add((-derive[1] + Math.Sqrt(discrim)) / (2 * derive[2]));
                        b.Add((-derive[1] - Math.Sqrt(discrim)) / (2 * derive[2]));
                    }

                    foreach (double s in b)
                    {
                        if (s < 1 && s > 0)
                        {
                            sol.Add(finalCurve[0] +
                                (finalCurve[1] * s) +
                                (finalCurve[2] * Math.Pow(s, 2)) +
                                (finalCurve[3] * Math.Pow(s, 3))
                                );
                        }
                    }
                }

                sol.Add(finalCurve[0]);
                sol.Add(finalCurve[0] + finalCurve[1] + finalCurve[2] + finalCurve[3]);

                Console.WriteLine(sol.Max() - sol.Min());
            }
        }
    }
}
